lib/corefoundation/base_wrapper.rb

```diff
@@ -0,0 +1,112 @@
+# The top level namespace for the corefoundation library. The raw FFI generated methods are attached here
+module CF
+  extend FFI::Library
+  ffi_lib "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation"
+
+  if FFI::Platform::ARCH == "x86_64"
+    typedef :long_long, :cfindex
+    typedef :long_long, :cfcomparisonresult
+    typedef :ulong_long, :cfoptionflags
+    typedef :ulong_long, :cftypeid
+    typedef :ulong_long, :cfhashcode
+  else
+    typedef :long, :cfindex
+    typedef :long, :cfcomparisonresult
+    typedef :ulong, :cfoptionflags
+    typedef :ulong, :cftypeid
+    typedef :ulong, :cfhashcode
+  end
+
+  typedef :pointer, :cftyperef
+
+  attach_function :show, "CFShow", [:cftyperef], :void
+  attach_function :release, "CFRelease", [:cftyperef], :void
+  attach_function :retain, "CFRetain", [:cftyperef], :cftyperef
+  attach_function "CFEqual", %i{cftyperef cftyperef}, :char
+  attach_function "CFHash", [:cftyperef], :cfhashcode
+  attach_function "CFCopyDescription", [:cftyperef], :cftyperef
+  attach_function "CFGetTypeID", [:cftyperef], :cftypeid
+
+  # @private
+  class Range < FFI::Struct
+    layout :location, :cfindex,
+           :length, :cfindex
+  end
+
+  # Provides a more declarative way to define all required abstract methods.
+  # This gets included in the CF::Base class whose subclasses then need to define those methods.
+  module BaseWrapper
+    @registered_types = {}
```

Comment:
Let's avoid keeping state here

Reply:
So I looked into this one after our discussion that day and tried out in a couple of ways. This seems to be the safest way to keep a single source of truth since the state doesn't get copied to any other instance/class and always stays as a part of the module accessible via `CF::BaseWrapper.instance_variables_get`

Reply:
can we try using `class instance variables` or `singleton`? Also, unless we plan to inherit/override the base class beyond 1 level (child), we can continue using the old design, i.e. class variable

Reply:
We can't use instance variables because then the state would be unique to each instance of class. Essentially the type checks would fail because we won't have the complete list of registered types at a single place. And while the singleton approach from old design did "sync" the value across multiple levels of subclasses, we always have a probability of someone overriding it if they are not careful. 

So essentially, the point is not about merely syncing the value, but also making sure that it's not accessible through a class or any of its subclasses. So in this case, `CF::Base.class_variables`, `CF::BaseSubClass.class_variables`, `CF::BaseSubSubClass.class_variables` all return empty set. But with a singleton approach they would all have a copy of the variable, thus allowing it to change via `CF::BaseSubSubClass.class_variables_set` which has CF::Base as its ancestor.

Reply:
We can't use instance variables because then the state would be unique to each instance of class. Essentially the type checks would fail because we won't have the complete list of registered types at a single place.

And while the singleton approach from old design did "sync" the value across multiple levels of subclasses, we always have a probability of someone overriding it if they are not careful. 

So essentially, the point is not about merely syncing the value, but also making sure that it's not accessible through a class or any of its subclasses. So in this case, `CF::Base.class_variables`, `CF::BaseSubClass.class_variables`, `CF::BaseSubSubClass.class_variables` all return empty set. But with a singleton approach they would all have a copy of the variable, thus allowing it to change via `CF::BaseSubSubClass.class_variables_set` which has `CF::Base` as its ancestor.

Reply:
IMO, we should stick with the `singleton` approach, that pattern is meant to solve the exact problem here.
Also, if subclasses have code like this, `CF::BaseSubSubClass.class_variables_set` there is no reason to assume they can't have code like, `BaseWrapper.instance_variable_set` and more.

Alright, with this, I am done with my review, you can follow up with others on my comments or we can schedule a call some time next week, involving Tim or Lamont or someone else with ruby expertise from infra team.

Reply:
Is this done?